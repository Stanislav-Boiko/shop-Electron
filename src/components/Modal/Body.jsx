
import React from 'react';
import './Body.scss';


function Body ({ children }) {
    return (

        <div className="text">{children}</div>

    )
};
  

export default Body;