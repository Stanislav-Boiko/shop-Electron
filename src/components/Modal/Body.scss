

.text {
    text-align: center;
    max-width: 450px;
    font-weight: 500;
    font-family: Arial, Helvetica, sans-serif;
    color: #6a0de4;
    line-height: 18.75px;
    font-size: 16px;
    margin: 0 auto 64px auto ;

    img {
        max-width: 180px;
        margin-bottom: 14px;
    }

    h3 {
        color: #6a0de4;
        font-size: 26px;
        margin: 10px 0;
    }
    p {
        color: #8a33fd;
        font-size: 16px;
        margin: 5px 0;
        font-weight: 600;
    }
}