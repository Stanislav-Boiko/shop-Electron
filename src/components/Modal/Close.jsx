
import React from 'react';
import './Close.scss';
import Button from '../Button/Button'; 


function Close({onClick}) {
    return (

        <Button children='X' classNames="closelImg" onClick={onClick}/>

    )
}


export default Close;