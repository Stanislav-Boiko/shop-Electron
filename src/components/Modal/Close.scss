


.closelImg {
    position: absolute;
    border: none;
    border-radius: 10px;
    color: white;
    background-color: rgba(138, 51, 253, 1);
    top: 20px;
    right: 20px;
    padding: 4px;
    width: 40px;
    height: 40px;
    cursor: pointer;
  
    &:hover {
        background-color: #6a0de4;
    }
}