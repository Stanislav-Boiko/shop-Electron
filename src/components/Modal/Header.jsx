import React from 'react';
import './Header.scss';


function Header ({ children }) {
    return (
        
        <h2 className="titleModal">{children}</h2>

    )
};


export default Header;