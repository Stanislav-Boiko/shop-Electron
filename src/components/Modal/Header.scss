


.titleModal {
    margin: 0 0 32px 0;
    font-family: Arial, Helvetica, sans-serif;
    font-size: 32px;
    font-weight: 400;
    line-height: 38.73px;
    color: #6a0de4;
  }