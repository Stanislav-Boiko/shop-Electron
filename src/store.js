import { configureStore, createSlice, createAsyncThunk } from "@reduxjs/toolkit";

// Асинхронний thunk для завантаження продуктів
export const fetchProducts = createAsyncThunk(
  "products/fetchProducts",
  async () => {
    const response = await fetch("/products.json");
    return await response.json();
  }
);

export const productsSlice = createSlice({
  name: "products",
  initialState: {
    items: [],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchProducts.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchProducts.fulfilled, (state, action) => {
        state.loading = false;
        state.items = action.payload;
      })
      .addCase(fetchProducts.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      });
  },
});

export const cartSlice = createSlice({
  name: "cart",
  initialState: {
    items: [], // масив id продуктів
  },
  reducers: {
    setCart: (state, action) => {
      state.items = action.payload;
    },
    addToCart: (state, action) => {
      if (!state.items.includes(action.payload)) {
        state.items.push(action.payload);
      }
    },
    removeFromCart: (state, action) => {
      state.items = state.items.filter(id => id !== action.payload);
    },
    clearCart: (state) => {
      state.items = [];
    }
  }
});

export const favoritesSlice = createSlice({
  name: "favorites",
  initialState: {
    items: [], // масив id продуктів
  },
  reducers: {
    setFavorites: (state, action) => {
      state.items = action.payload;
    },
    addToFavorites: (state, action) => {
      if (!state.items.includes(action.payload)) {
        state.items.push(action.payload);
      }
    },
    removeFromFavorites: (state, action) => {
      state.items = state.items.filter(id => id !== action.payload);
    },
    clearFavorites: (state) => {
      state.items = [];
    }
  }
});

export const modalSlice = createSlice({
  name: "modal",
  initialState: {
    isOpen: false,
    title: "",
    text: "",
    footerProps: {},
    children: null,
  },
  reducers: {
    openModal: (state, action) => {
      return { ...state, ...action.payload, isOpen: true };
    },
    closeModal: (state) => {
      return { ...state, isOpen: false };
    },
  },
});



export const { setCart, addToCart, removeFromCart, clearCart } = cartSlice.actions;
export const { setFavorites, addToFavorites, removeFromFavorites, clearFavorites } = favoritesSlice.actions;
export const { openModal, closeModal } = modalSlice.actions;

export const store = configureStore({
  reducer: {
    products: productsSlice.reducer,
    modal: modalSlice.reducer,
    cart: cartSlice.reducer,
    favorites: favoritesSlice.reducer,
  },
});