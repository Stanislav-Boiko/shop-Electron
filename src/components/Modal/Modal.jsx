import React from 'react';
import Wrapper from './Wrapper';
import Close from './Close';
import Header from './Header';
import Body from './Body';
// import Footer from '..Button/ButtonContainer';
import Footer from '../Button/ButtonContainer'
import { useDispatch } from 'react-redux';
import { addToCart, removeFromCart, removeFromFavorites, closeModal } from '../../store';

function Modal({ isOpen, onClose, title, product, actionType, actionPayload }) {
  const dispatch = useDispatch();

  const handleConfirm = () => {
    if (actionType === "addToCart") {
      dispatch(addToCart(actionPayload));
    }
    if (actionType === "removeFromCart") {
      dispatch(removeFromCart(actionPayload));
    }
    if (actionType === "removeFromFavorites") {
      dispatch(removeFromFavorites(actionPayload));
    }
    dispatch(closeModal());
  };

  if (!isOpen) return null;

  return (
    <Wrapper onClose={onClose}>
      <Close onClick={onClose}/>
      <Header>{title}</Header>
      {product && (
        <Body>
          <div>
            <img src={product.image} alt={product.name} />
            <h3>{product.name}</h3>
            <p>Color: {product.color}</p>
            <p>Sku: {product.sku}</p>
            <p>Price: ${product.price}</p>
          </div>
        </Body>
      )}
      <Footer
        firstText="NO, CANCEL"
        firstClassNames="white-btn"
        firstClick={onClose}
        secondaryText="YES, CONFIRM"
        secondaryClassNames="violet-btn"
        secondaryClick={handleConfirm}
      />
    </Wrapper>
  );
}

export default Modal;

