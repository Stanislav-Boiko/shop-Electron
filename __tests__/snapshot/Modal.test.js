import React from 'react';
import { render } from '@testing-library/react';
import Modal from 'components/Modal/Modal';
import { Provider } from 'react-redux';
import { store } from 'store';

test('Modal не рендериться, якщо isOpen=false', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <Modal
        isOpen={false}
        onClose={() => {}}
        title="Test Modal"
        product={null}
        actionType=""
        actionPayload=""
      />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('Modal рендериться з продуктом', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <Modal
        isOpen={true}
        onClose={() => {}}
        title="Test Modal"
        product={{ name: "Test", color: "Red", sku: "123", price: 100, image: "/test.png" }}
        actionType="addToCart"
        actionPayload="123"
      />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

