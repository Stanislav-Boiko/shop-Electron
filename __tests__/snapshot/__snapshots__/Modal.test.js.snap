// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal не рендериться, якщо isOpen=false 1`] = `<DocumentFragment />`;

exports[`Modal рендериться з продуктом 1`] = `
<DocumentFragment>
  <div
    class="overlay"
  >
    <div
      class="modal"
    >
      <button
        class="closelImg"
        type="button"
      >
        X
      </button>
      <h2
        class="titleModal"
      >
        Test Modal
      </h2>
      <div
        class="text"
      >
        <div>
          <img
            alt="Test"
            src="/test.png"
          />
          <h3>
            Test
          </h3>
          <p>
            Color: Red
          </p>
          <p>
            Sku: 123
          </p>
          <p>
            Price: $100
          </p>
        </div>
      </div>
      <div
        class="buttons-container"
      >
        <button
          class="white-btn"
          type="button"
        >
          NO, CANCEL
        </button>
        <button
          class="violet-btn"
          type="button"
        >
          YES, CONFIRM
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
